src/nano_contracts/deserializer.ts

```diff
@@ -192,53 +257,85 @@ class Deserializer {
-    // eslint-disable-next-line prefer-const -- To split this declaration would be confusing
-    [size, signedBuffer] = unpackToInt(2, false, signedData);
-    let parsed = this.deserializeFromType(signedBuffer.slice(0, size), valueType);
-    if (valueType === 'bytes') {
+    // Reading argument
+    const parseResult = this.deserializeFromType(signedData, internalType);
+    let parsed = parseResult.value;
+    const bytesReadFromValue = parseResult.bytesRead;
+
+    if (internalType === 'bytes') {
       // If the value is bytes, we should transform into hex to return the string
-      parsed = (parsed as Buffer).toString('hex');
+      parsed = bufferToHex(parsed as Buffer);
+    }
+
+    if (internalType === 'bool') {
+      parsed = (parsed as boolean) ? 'true' : 'false';
     }
-    signedBuffer = signedBuffer.slice(size);
-    return `${bufferToHex(signedBuffer)},${parsed},${valueType}`;
+
+    // Reading signature
+    const { value: parsedSignature, bytesRead: bytesReadFromSignature } = this.deserializeFromType(
+      signedData.subarray(bytesReadFromValue),
+      'bytes'
+    );
+
+    return {
+      value: `${bufferToHex(parsedSignature as Buffer)},${parsed},${internalType}`,
+      bytesRead: bytesReadFromValue + bytesReadFromSignature,
+    };
   }
 
   /**
    * Deserialize a value decoded in bytes to a base58 string
    *
-   * @param {value} Value to deserialize
+   * @param {Buffer} buf Value to deserialize
    *
    * @memberof Deserializer
    * @inner
    */
-  toAddress(value: Buffer): string {
+  toAddress(buf: Buffer): DeserializeResult<string> {
+    const lenReadResult = leb128Util.decodeUnsigned(buf, 1);
+    if (lenReadResult.value !== 25n) {
+      // Address should be exactly 25 bytes long
+      throw new Error('Address should be 25 bytes long');
+    }
     // First we get the 20 bytes of the address without the version byte and checksum
-    const addressBytes = value.slice(1, 21);
+    const addressBytes = buf.subarray(2, 22);
     const address = helpersUtils.encodeAddress(addressBytes, this.network);
+    address.validateAddress();
     const decoded = address.decode();
-    if (decoded[0] !== value[0]) {
+    if (decoded[0] !== buf[1]) {
+      throw new Error(
+        `Asked to deserialize an address with version byte ${buf[0]} but the network from the deserializer object has version byte ${decoded[0]}.`
+      );
+    }
+    if (decoded.subarray(21, 25).toString('hex') !== buf.subarray(22, 26).toString('hex')) {
+      // Checksum value generated does not match value from fullnode
       throw new Error(
-        `Asked to deserialize an address with version byte ${value[0]} but the network from the deserializer object has version byte ${decoded[0]}.`
+        `When parsing and Address(${address.base58}) we calculated checksum(${decoded.subarray(21, 25).toString('hex')}) but it does not match the checksum it came with ${buf.subarray(22, 26).toString('hex')}.`
```

Comment:
suggestion(non-blocking): Create variables for storing both checksums, so that it's harder to accidentally add an issue in the error message like the one that happened above.

Reply:
Done